hedera-node/docs/design/services/smart-contract-service/system-contract-versioning.md

```diff
@@ -0,0 +1,82 @@
+# System Contract Versioning
+
+## Purpose
+
+This document explores the implementation changes needed to support and test versioning of system contracts in the Hedera Smart Contract Service.
+
+## Architecture and Implementation
+
+### Relevant Current Implementation
+
+Currently, the handling of HTS, HAS and HSS system function calls follows similar pattern as described below:
+
+1. A particular descendant of the `SystemContract` class is identified by the contract address for processing the call.
+2. A descendant of the `AbstractCallAttempt` class is created from the call data and other information available in the message frame.
+3. The `AbstractCallAttempt` descendant is tested using the `identifyMethod` against a list of class that implement the `CallTranslator` interface provided by dagger in order to determine which translator is responsible for processing the function call.
+4. The `CallTranslator` descendant decodes and calls the necessary helper classes to either dispatch or process the call.
+5. The `CallTranslator` also keeps metadata about the system contract function in the `SystemContractMethod` class.
+
+### Proposed Changes
+
+The following changes are needed in order to support versioning of system contracts:
+1. Descendants of the `SystemContract` class will map to more than one contract address.  The dagger provided map allows for the mapping of multiple address to a single `SystemContract` class.
+2. Descendants of the `AbstractCallAttempt` class will be augmented with a system contract address field.  Otherwise the current common `AbstractCallAttempt` and related `CallFactory` class descendants will be used commonly by all system contracts of the same type.
+3. The `SystemContractMethod` class metadata will be augmented with an enum set of supported system contract addresses which can include an item that can denote _all_ addresses.
+4. The `isMethod` function support by `CallTranslator` descendants will be augmented to check the system contract address field in the `AbstractCallAttempt` descendant in order to determine a matching translator.
+
+### Package Organization
+
+In order to organize and support multiple addresses for a system contract, a sensible package must be introduced.
+Currently , all `CallTranslator` descendant classes and related helper classes are packaged together under the system contract package.
+Functions supported under a new system contract address for the same system contract type will be placed in a new package with the name `address_<system_contract_address>`.
+The packages at top level for a system contract will contain the base implementation which will be used if no override implementation of the function is available.
+
+For example, the HTS system contract will have the following packages where a new address for the HTS system contract at address `0x16c` is introduced and overrides the `create` function:
```

Comment:
Alternate organization (of course) would be to put each alternate method under the method "topic", e.g., something like
```
|--- hts
   airdrops/
      <some-contract-for-airdrops-at-the-original-address>.sol/
      <another-contract-for-airdrops-atthe-original-address>.sol/
      0x16c/
         <a-contract-for-airdrops-at-0x16c>.sol/
```

So were there pros/cons that led you to choose the organization you choose:

e.g.,
groups all methods for a given address together vs groups all methods for a topic together but you can't as easily find all methods for a given address
etc.